internal/gcs/container.go

```diff
@@ -225,20 +245,83 @@ func (c *Container) Terminate(ctx context.Context) (err error) {
 // Wait waits for the container to terminate (or Close to be called, or the
 // guest connection to terminate).
 func (c *Container) Wait() error {
+	<-c.waitBlock
+	return c.waitError
+}
+
+// waitBackground waits for the GuestConnection to notify of container exit and
+// unblocks other Wait calls
+//
+// This MUST be called via a goroutine to wait on a background thread, and can
+// only be called once
+func (c *Container) waitBackground(ctx context.Context) {
+	var err error
+	ctx, span := trace.StartSpan(ctx, "gcs::Container::waitBackground")
+	defer span.End()
+	defer func() { oc.SetSpanStatus(span, err) }()
+	span.AddAttributes(trace.StringAttribute(logfields.ContainerID, c.id))
+
+	// wait for container exit
 	select {
-	case <-c.notifyCh:
-		return nil
 	case <-c.closeCh:
-		return errors.New("container closed")
+		// container in the guest was closed
+	case <-c.waitBlock:
+		// `c.Close()` was called
 	}
+
+	c.closeOnce.Do(func() {
+		// `notifyCh` and `closeCh` are closed by `GuestConnection`, so do not close here
+		close(c.waitBlock)
+		c.notifications.Close()
+	})
```

Comment:
could we just call c.Close here?

Reply:
`c.Close` also sets a `waitErr` field to show that the container wasn't exited, but the connection to it was closed. I modeled this after [`hcs.System`](https://github.com/microsoft/hcsshim/blob/master/internal/hcs/system.go#L254) and [`jobcontainers.JobContainers`](https://github.com/microsoft/hcsshim/blob/master/internal/jobcontainers/jobcontainer.go#L491), where `Wait` shows the container was closed but not shut down.